src/core/cgroup.c

```diff
@@ -4178,6 +4178,16 @@ int unit_cgroup_freezer_action(Unit *u, FreezerAction action) {
         if (!u->cgroup_realized)
                 return -EBUSY;
 
+        if (action == FREEZER_THAW) {
+                Unit *slice = UNIT_GET_SLICE(u);
+
+                if (slice && !unit_has_name(slice, SPECIAL_ROOT_SLICE)) {
```

Comment:
hmmm, I think the unit_has_name() should actually move into the top of unit_cgroup_freezer_action(), i.e. we should refuse all freeze operations on the root slice early on (i.e. fail with an error). and we should make all thaw operations on the root slice a NOP (i.e. succeed but not do anything).

And in fact, we should do the same for SPECIAL_INIT_SCOPE so that we don't end up freezing ourselves.